### PR TITLE
feat: add avg pooling to embedding VGGishAudioEncoder

### PR DESCRIPTION
**Pull request by winstonww**
This PR adds average pooling to as the final step to calculating embedding in VGGishAudioEncoder. We need to take the mean of the examples in the doc/chunk because computing cosine similarity currently requires the two objects to be 1D vectors. (For details, please see the VGGishAudioEncoder in eap-qupeiyin). Also adding the functionality to download model if model is not found. Thank you.

### CODE REVIEW

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 1]
```diff
@@ -33,6 +35,9 @@ def __init__(self,
         self.pca_path = pca_path
         self.default_traversal_paths = default_traversal_paths or ['r']
 
+        if not os.path.exists(self.model_path) or not os.path.exists(self.pca_path):
+            subprocess.call('scripts/download_model.sh')
+
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:40]: please avoid to use `subprocess`, use `requests` to download the model instead.

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/unit/test_exec.py [hunk 2]
```diff
@@ -25,6 +25,7 @@ def test_embedding_dimension():
     ops.reset_default_graph()
     model = VggishAudioEncoder()
     model.encode(doc, parameters={})
+    assert doc[0].embedding.shape[0] == 128
     assert doc[0].embedding.shape[-1] == 128
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/tests/unit/test_exec.py:29]: so the embedding shape should be `doc[0].embedding.shape == (128,0)`?

File: jinahub/encoders/audio/VGGISHAudioEncoder/requirements.txt [hunk 3]
```diff
@@ -1,7 +1,7 @@
 jina-commons @ git+https://github.com/jina-ai/jina-commons.git#egg=jina-commons
 jina[hub]==2.0.7
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/requirements.txt:2]: remove `jina` depedency from requirements.txt

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jinahub/encoders/audio/VGGISHAudioEncoder/requirements.txt (hunk 3 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/requirements.txt:2]: Addressed. Please see [here](https://github.com/jina-ai/executors/pull/31/commits/adc062852447ab0fa23e4b5dbc3cbcdba9e083ff) for the change, thank you.

File: jinahub/encoders/audio/VGGISHAudioEncoder/tests/unit/test_exec.py (hunk 2 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/tests/unit/test_exec.py]: Yes, I have updated the code to assert the above. Please see [here](https://github.com/jina-ai/executors/pull/31/commits/adc062852447ab0fa23e4b5dbc3cbcdba9e083ff) for the change.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 1 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py]: Addressed, please see [here](https://github.com/jina-ai/executors/pull/31/commits/adc062852447ab0fa23e4b5dbc3cbcdba9e083ff) for the change.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 4]
```diff
@@ -33,6 +36,19 @@ def __init__(self,
...
+        if not os.path.exists(self.model_path):
+            r = requests_.get('https://storage.googleapis.com/audioset/vggish_model.ckpt')
+            with open(model_path, 'wb') as f:
+                f.write(r.content)
+
+        if not os.path.exists(self.pca_path):
+            r = requests_.get('https://storage.googleapis.com/audioset/vggish_pca_params.npz')
+            with open(pca_path, 'wb') as f:
+                f.write(r.content)
+
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:50]: need to refactor executor's parameters. We should use the same folder (default is `.models`) for vggish and pca model. 
```
self.model_path = Path(model_path)
self.vgg_model_path = self.model_path / 'vggish_model.ckpt'
self.pca_model_path = self.model_path / 'vggish_pca_params.npz'
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 4 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py]: Addressed, please see [here](https://github.com/jina-ai/executors/pull/31/commits/c8d0239634a96d11d2f0b684c83a65ab6b5f78d0) for the change. Also added error checking when issuing requests to download models as discussed, thank you.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 5]
```diff
@@ -2,11 +2,15 @@
...
+import requests as requests_
 from typing import Any, Optional, List, Iterable
 import tensorflow as tf
 tf.compat.v1.disable_eager_execution()
 
 from jina import Executor, requests, DocumentArray
+from jina.logging.logger import JinaLogger
 from vggish.vggish_postprocess import *
 from vggish.vggish_slim import *
```
Comment (bwanglzu) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:15]: can we reorder the imports a bit? 
```python
# os dependencies
import os

# 3rd party dependencies
import requests as requests_
...

# library dependencies
from .helper import helper_func
...
```

**Review by bwanglzu — COMMENTED** [review]
minor comment, great work!

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 5 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:16]: I have reordered the imports [here](https://github.com/jina-ai/executors/pull/31/commits/d3786b57f0d6474ad54d7060df4f351f0ce62c44). Let me know if this looks okay to you, thank you.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 6]
```diff
@@ -2,11 +2,16 @@
 __license__ = "Apache-2.0"
 
 import os
+from pathlib import Path
 from typing import Any, Optional, List, Iterable
+
+from jina import Executor, requests, DocumentArray
+from jina.logging.logger import JinaLogger
+import requests as requests_
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:10]:
```suggestion
import requests as _requests
```

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 7]
```diff
@@ -17,30 +22,63 @@ class VggishAudioEncoder(Executor):
...
-     :param model_path: path of the pre-trained Vggish model
-     :param pca_path: path of the pre-trained PCA model
+     :param model_path: path of the models directory
      :param default_traversal_paths: fallback batch size in case there is not batch size sent in the request
      """
 
     def __init__(self,
-                 model_path: str = os.path.join(cur_dir, 'models/vggish_model.ckpt'),
-                 pca_path: str = os.path.join(cur_dir, 'models/vggish_pca_params.npz'),
+                 model_path: str = os.path.join(cur_dir, 'models'),
```
Comment (numb3r3) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:30]:
```suggestion
                 model_path: str = cur_dir / 'models'
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 6 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py]: Addressed, please see [here](https://github.com/jina-ai/executors/pull/31/commits/d5752827dc4e83eb858cb90d75d24d2b01830e0e) for the change.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 7 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py]: Addressed, please see [here](https://github.com/jina-ai/executors/pull/31/commits/d5752827dc4e83eb858cb90d75d24d2b01830e0e) for the change, thank you.

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py [hunk 8]
```diff
@@ -70,14 +108,15 @@ def _get_input_data(self, docs: DocumentArray, parameters: dict):
...
+
     def _create_embeddings(self, filtered_docs: Iterable):
         """Update the documents with the embeddings generated by VGGISH"""
 
         for d in filtered_docs:
             # Vggish broadcasts across different length audios, not batches
             [embedding] = self.sess.run([self.embedding_tensor], feed_dict={self.feature_tensor: d.blob})
             result = self.post_processor.postprocess(embedding)
-            d.embedding = (np.float32(result) - 128.) / 128.
+            d.embedding = np.mean((np.float32(result) - 128.) / 128., axis=0)
```
Comment (Yongxuanzhang) [inline jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:119]: is it possible we need other like max pooling methods?

**Review by numb3r3 — APPROVED** [review]
LGTM

File: jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py (hunk 8 above)
Comment (winstonww) [reply jinahub/encoders/audio/VGGISHAudioEncoder/vggish_audio_encoder.py:119]: Thanks for the suggestion, for eap-qupeiyin, we are using average pooling. To add other options such as max pooling, I think we can leave this for the next PR.

**Pull request merged**